src/linker/Linker/Annotations.cs

```diff
@@ -161,15 +161,16 @@ public bool HasSubstitutedInit (TypeDefinition type)
 		[Obsolete ("Mark token providers with a reason instead.")]
 		public void Mark (IMetadataTokenProvider provider)
 		{
+			// No origin provided, so use the provider itself if possible
 			if (!processed.Contains (provider))
-				marked_pending.Add (provider);
+				marked_pending.TryAdd (provider, new MessageOrigin (provider as ICustomAttributeProvider));
```

Comment:
Why'd this change to `TryAdd` - was `Add` causing problems? If so I wonder if it might be hiding a bug.

Reply:
`Add` throws if the value is already in the dictionary (unlike `HashSet`). So the `TryAdd` is there to allow calling `Mark` on the same thing multiple times (which does happen quite a bit).
I don't think it's a bug, maybe inefficiency... hard to tell.

Reply:
Sorry, I forgot that `HashSet.Add` doesn't throw. Makes sense.